### DISCUSSION TITLE
Allow custom search param parser to be passed to "useSearch"

### DISCUSSION THREAD

**Discussion by briavicenti** [TanStack/router#553] [Ideas]
I currently am passing a numerical ID string as a search param, which then gets parsed as a number with the default `parseSearch` behavior:
```
navigate({
  search: {
    id: "0000",
  },
});
```
As you can imagine, this doesn't work for 0-prefixed ids such as the one above, because they are set to the equivalent numerical primitive (`0000` --> `0`, `0012` --> `12`, etc).

To get around this, I've implemented the following `parseSearch` method that uses `URLSearchParams` to parse all param values as strings:
```
const reactLocation = new ReactLocation({
  parseSearch: (searchStr) => {
    const searchParams = new URLSearchParams(searchStr);
    return Object.fromEntries(searchParams.entries());
  },
});
```
However, the fact that this parsing behavior is app-wide will make it onerous to parse actual numerical or boolean primitives on other screens where we are using search params. I've also yet to figure out a way to override the React Location types such that `MakeGenerics` does not accept an object with non-string values, meaning that devs in my repo may type a param as a `number` but actually receive a `string` instead if they don't notice this top-level parsing override.

I would _love_ to implement a custom parser only for this sole case where we are sure we need one. Would it be possible to add a param for the `useSearch` hook for a custom parser that overrides the top-level default?

**Comment by erengy** [comment]
I ended up quoting the value (e.g. `id: '"0012"'`) as a workaround. `defaultParseSearch` calls `JSON.parse`, which then returns a string without the quotes.
